drivers/adc/adc_stm32.c

```diff
@@ -1174,6 +1176,60 @@ static const struct adc_driver_api api_stm32_driver_api = {
 	.ref_internal = STM32_ADC_VREF_MV, /* VREF is usually connected to VDD */
 };
 
+/*
+ * Prefix default when not configured by the DTS
+ * depending on the serie (stm32F1 has none)
+ */
+#if defined(CONFIG_SOC_SERIES_STM32L1X) || \
+	defined(CONFIG_SOC_SERIES_STM32U5X)
+#define prescaler_default	LL_ADC_CLOCK_ASYNC_DIV
+#elif !defined(CONFIG_SOC_SERIES_STM32F1X)
+#define prescaler_default	LL_ADC_CLOCK_SYNC_PCLK_DIV
+#endif
+
+/* st_prescaler property requires 2 elements : clock ASYNC/SYNC and DIV */
+#define STM32_ADC_CLOCK(x, prop)	DT_INST_PROP_BY_IDX(x, prop, 0)
+#define STM32_ADC_DIV(x, prop)		DT_INST_PROP_BY_IDX(x, prop, 1)
+
+/*
+ * Macro to check the property is valid with two elements:
+ * 1st one is SYNC or ASYNC (not all series have both)
+ * 2nd one from 1 to 256 (not all series have full range)
+ */
+#define ADC_STM32_HAS_PROP(x, prop)					\
+	IF_ENABLED(DT_INST_NODE_HAS_PROP(x, prop),			\
+		(COND_CODE_1(						\
+			IS_EQ(DT_INST_PROP_LEN(x, prop), 2),		\
+			(BUILD_ASSERT((					\
+				((STM32_ADC_CLOCK(x, prop) == SYNC) ||	\
+				(STM32_ADC_CLOCK(x, prop) == ASYNC)) &&	\
+				((STM32_ADC_DIV(x, prop) >= 1) &&	\
+				(STM32_ADC_DIV(x, prop) <= 256))),	\
+				"Invalid ADC "#prop" property")),	\
+			(BUILD_ASSERT((DT_INST_PROP_LEN(x, prop) == 2),	\
+				"Invalid ADC "#prop" property length")))))
```

Comment:
Since there is no real reason to use COND_CODE logic here I wonder if this could be simplified with the following :
```suggestion
IF_ENABLED(DT_INST_NODE_HAS_PROP(x, prop), \
			(BUILD_ASSERT(FOO) || (BUILD_ASSERT(BAR))
```

Reply:
ORing BUILD_ASSERT in that way, will give a compilation error in on or the other case:
```
error: expected ';' before '||' token
 1218 |    || BUILD_ASSERT 
```